pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,276 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package interruption
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	sqsapi "github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/apis"
+	"github.com/aws/karpenter/pkg/apis/config/settings"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/cache"
+	interruptionevents "github.com/aws/karpenter/pkg/controllers/interruption/events"
+	"github.com/aws/karpenter/pkg/controllers/interruption/messages"
+	"github.com/aws/karpenter/pkg/controllers/interruption/messages/statechange"
+	"github.com/aws/karpenter/pkg/controllers/providers"
+	"github.com/aws/karpenter/pkg/errors"
+	"github.com/aws/karpenter/pkg/utils"
+
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/operator/scheme"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+func init() {
+	lo.Must0(apis.AddToScheme(scheme.Scheme))
+}
+
+// Controller is an AWS interruption controller.
+// It continually polls an SQS queue for events from aws.ec2 and aws.health that
+// trigger node health events or node spot interruption/rebalance events.
+type Controller struct {
+	kubeClient                client.Client
+	clk                       clock.Clock
+	recorder                  events.Recorder
+	sqsProvider               *providers.SQS
+	unavailableOfferingsCache *cache.UnavailableOfferings
+	parser                    *EventParser
+}
+
+func NewController(kubeClient client.Client, clk clock.Clock, recorder events.Recorder,
+	sqsProvider *providers.SQS, unavailableOfferingsCache *cache.UnavailableOfferings) *Controller {
+
+	return &Controller{
+		kubeClient:                kubeClient,
+		clk:                       clk,
+		recorder:                  recorder,
+		sqsProvider:               sqsProvider,
+		unavailableOfferingsCache: unavailableOfferingsCache,
+		parser:                    NewEventParser(DefaultParsers...),
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	if settings.FromContext(ctx).EnableInterruptionHandling {
```

Comment:
Consider using a shortcircuitting pattern to minimize nesting.
```suggestion
	if !settings.FromContext(ctx).EnableInterruptionHandling {
	   return reconcile.Result{}, nil
	}
```